File: py/day03.py
# %%
import numpy as np

# %%
with open('../data/day03.txt', 'r') as f:
    data = np.array([list(map(int, list(line))) for line in f.read().splitlines()], dtype=int)
# eps = (np.sign((data * 2 - 1).sum(0)) + 1) // 2
# delta = 1 - eps

# eps = int(''.join(list(map(str, eps))), 2)
# delta = int(''.join(list(map(str, delta))), 2)
# eps * delta
# %%
i = 0
while len(data) > 1 and i < data.shape[1]:
    v, c = np.unique(data[:, i], return_counts=True)
    print(v, c)
    bit = v[np.argmin(c)]
    if c[0] == c[1]:
        bit = 0
    data = data[data[:, i] == bit]
    i += 1
print(data)
print(int(''.join(list(map(str, data[0]))), 2))

# %%
# 1059 * 2808

File: rs/day03.rs
fn majority_elmt(v: &Vec<u32>) -> Option<u32> {
    let mut m = 0;
    let mut i = 0;
    for elmt in v {
        if i == 0 {
            m = *elmt;
            i += 1;
        } else if m == *elmt {
            i += 1;
        } else {
            i -= 1;
        }
    }
    match i {
        0 => None,
        _ => Some(m),
    }
}

fn main() {
    let data: Vec<Vec<u32>> = include_str!("../data/day03.txt")
        .lines()
        .map(|line| line.chars().map(|x| x.to_digit(10).unwrap()).collect())
        .collect();


    let n = data[0].len() as u32;
    let mut eps = 0;
    let mut delta = 0;
    for i in 0..n {
        let col: Vec<u32> = data.iter().map(|v| v[i as usize]).collect();
        let bit = match majority_elmt(&col) {
            Some(m) => m,
            None => 1,
        };
        eps += bit * 2_u32.pow(n - i - 1);
        delta += (1 - bit) * 2_u32.pow(n - i - 1);
    }
    println!("{:?}", eps * delta);

}